### PR TITLE
Preserve float64 dense xent tail gradients

### LINKED ISSUE #126634: tf.nn.softmax_cross_entropy_with_logits loses a finite first-derivative signal for float64 input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.softmax_cross_entropy_with_logits` returns zero for one logit-gradient component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

Softmax cross entropy is invariant to adding the same constant to all logits, so the gradient with respect to logits should sum to zero. In the reproducer below, the expected gradient is approximately `[-5.551115123125776e-17, 5.551115123125776e-17]`.

However, TensorFlow returns `[[0.0, 5.551115123125776e-17]]`. The first gradient component is incorrectly collapsed to zero, and the resulting logit gradient no longer preserves the expected zero-sum relation.

Expected behavior?

The first-order autodiff result for `tf.nn.softmax_cross_entropy_with_logits` should preserve the finite, normal, nonzero `float64` gradient component. For the given logits and labels, the gradient should be close to `[[-5.551115123125776e-17, 5.551115123125776e-17]]`, and the gradient sum should be zero.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

dtype = tf.float64

expected = [-5.551115123125776e-17, 5.551115123125776e-17]

logits = tf.Variable([[37.42994775023705, 0.0]], dtype=dtype)
labels = tf.constant([[1.0, 0.0]], dtype=dtype)

with tf.GradientTape() as tape:
    loss = tf.nn.softmax_cross_entropy_with_logits(labels=labels, logits=logits)
    objective = tf.reduce_sum(loss)

actual = tape.gradient(objective, logits)
grad_sum = tf.reduce_sum(actual)

print("loss:")
print(loss.numpy().tolist())
print("actual:")
print(actual.numpy().tolist())
print("expected:")
print(expected)
print("gradient sum:")
print(float(grad_sum))

if actual.numpy()[0, 0] == 0.0 and expected[0] != 0.0:
    print("BUG REPRODUCED: tf.nn.softmax_cross_entropy_with_logits loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
loss:
[0.0]
actual:
[[0.0, 5.551115123125776e-17]]
expected:
[-5.551115123125776e-17, 5.551115123125776e-17]
gradient sum:
5.551115123125776e-17
BUG REPRODUCED: tf.nn.softmax_cross_entropy_with_logits loses a finite first-derivative signal
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
## Summary

- retain the normalized float64 probabilities while forming the dense cross-entropy gradient
- remove the row-sum rounding residual along the probability vector, preserving finite tail components and translation invariance
- leave lower-precision execution unchanged

Fixes #126634.

## Testing

- `bazel build //tensorflow/core/kernels:xent_op`
- added graph/eager regression coverage for the reported float64 input

### CODE REVIEW

**Comment by GodlyDonuts** [comment]
@gemini-code-assist /review

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for the PR to preserve float64 dense xent tail gradients!

The GitHub CI tests failed on both compilation jobs (CPU and CUDA) as well as the PyLint check. 

1. **Compilation Failures**: It looks like `is_same_v` and `if constexpr` were introduced in `xent_op.h` which causes compilation errors on some of the CI toolchains. Could you please fallback to `std::is_same<T, double>::value` and avoid `if constexpr` to ensure it successfully builds across all CI compilers?

2. **PyLint Failure**: PyLint failed on `tensorflow/python/kernel_tests/nn_ops/xent_op_test_base.py` with:
   `xent_op_test_base.py:54:8: E1130: bad operand type for unary -: NoneType (invalid-unary-operand-type)`
   Could you also fix this lint error in the file while you're at it?

Looking forward to the updates!

**Review by dmiltr3 — APPROVED** [review]
Thank you for the PR! The changes solidly address the float64 xent bugs by properly preserving the finite tail gradients of softmax outputs.

I have verified the math logic, resolved the vendored google3 deps, and ran this through our full internal CI suite (TAP). Everything passes perfectly.

LGTM! I will now proceed to authorize this PR for final merge.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for the update! 

During internal automated testing, `xent_op_test` failed to build due to a missing strict dependency in Bazel:
```
PyStrictDepsCheck returned a non-zero exit code:
Target 'xent_op_test_base' missing dependency for imported module 'tensorflow.python.ops.gen_nn_ops'.
```

In `tensorflow/python/kernel_tests/nn_ops/xent_op_test_base.py`, `gen_nn_ops` was imported for the new test case, but `"//tensorflow/python/ops:nn_ops_gen"` is missing from the `deps` list of the `xent_op_test_base` target in `tensorflow/python/kernel_tests/nn_ops/BUILD`.

Could you please add `"//tensorflow/python/ops:nn_ops_gen"` to `deps` of `xent_op_test_base` in `tensorflow/python/kernel_tests/nn_ops/BUILD`?

```diff
diff --git a/tensorflow/python/kernel_tests/nn_ops/BUILD b/tensorflow/python/kernel_tests/nn_ops/BUILD
--- a/tensorflow/python/kernel_tests/nn_ops/BUILD
+++ b/tensorflow/python/kernel_tests/nn_ops/BUILD
@@ -936,6 +936,7 @@ py_library(
         "//tensorflow/python/ops:gradients_impl",
         "//tensorflow/python/ops:math_ops",
         "//tensorflow/python/ops:nn_grad",
         "//tensorflow/python/ops:nn_ops",
+        "//tensorflow/python/ops:nn_ops_gen",
         "//tensorflow/python/platform:client_testlib",
         "@absl_py//absl/testing:parameterized",
     ],
```

Once this dependency is added, the test target will build cleanly and pass the automated test suite. Thank you!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi, thanks for the PR! 

While reviewing and running our internal test suites, the newly added `testDoublePreservesSmallGradient` test failed on our internal CPU test builds which utilize specialized CPU kernels (e.g. OneDNN/MKL). 

It appears that the fix in `xent_op.h` successfully updates the fallback Eigen C++ kernel, but does not cover the specialized CPU kernels that TensorFlow defaults to on certain architectures. This results in the test failing with gradients evaluating to `0.0` instead of the expected tail probability:

```
Mismatch at index:
 [0, 0]: -5.551115123125776e-17 (ACTUAL), 0.0 (DESIRED)
 ACTUAL: array([[-5.551115e-17,  5.551115e-17]])
 DESIRED: array([[0.000000e+00, 5.551115e-17]])
```
*(Note: test expected `[[-t, t]]` but the kernel returned `[[0.0, t]]`)*

Could you please investigate updating the corresponding optimized CPU kernels (like OneDNN) to ensure the tail components and translation invariance are preserved there as well, so that the tests pass consistently across all CPU backends?

Let me know if you have any questions or need help navigating the CPU kernels!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hello! Thank you for the pull request. We have pulled your changes in and started running internal tests.

During our internal presubmit process on Google's infrastructure, we discovered that `testDoublePreservesSmallGradient` is failing with a gradient of `0.0`. Upon investigation, we found a subtle Eigen aliasing issue in `XentGradientCorrector<CPUDevice, double>::Correct` in `xent_op.cc`. 

Specifically, in this code:
```cpp
auto probabilities = backprop + labels.broadcast(labels_bcast);
scratch.reshape(batch_only).device(d) = backprop.sum(along_class);
// ...
backprop.device(d) =
    backprop - scratch.broadcast(one_by_class) * probabilities;
```
Because `probabilities` captures `backprop` as a reference, lazy evaluation during the inplace `backprop.device(d) = ...` assignment evaluates `probabilities` using partially updated or chunk-wise modified values of `backprop` under certain compiler optimizations (like the ones used in our internal builds). This causes the corrector to silently drop the small gradient, resulting in exactly `0.0` instead of `-5.55e-17`.

To fix this, we need to explicitly force evaluation into a temporary tensor by appending `.eval()` to the expression:

```cpp
auto probabilities = (backprop + labels.broadcast(labels_bcast)).eval();
```

I have verified locally that this explicitly breaks the aliasing and correctly preserves the tail gradients, allowing `testDoublePreservesSmallGradient` to successfully pass. 

Could you please add `.eval()` to `probabilities` in `xent_op.cc` and push the update? 

Once updated, we will rerun the internal presubmit and get this merged for you! Thank you again.

**Review by dmiltr3 — APPROVED** [review]
All internal presubmit tests have passed successfully. The changes look good and resolve the previous failures.

LGTM. Thank you for the contribution!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @GodlyDonuts,

Thank you for your pull request!

Our automated presubmit test suite encountered an issue when validating the build targets for this change. Specifically, strict dependency checking failed on `//tensorflow/python/kernel_tests/nn_ops:xent_op_test_base`:

### Issue
In `tensorflow/python/kernel_tests/nn_ops/xent_op_test_base.py`, `gen_nn_ops` is directly imported:
```python
from tensorflow.python.ops import gen_nn_ops
```
and `tensorflow/python/kernel_tests/nn_ops/BUILD` added:
```python
"//tensorflow/python/ops:nn_ops_gen",
```
However, in the internal build rules, `nn_ops_gen` is a private op generator target, and direct imports of low-level generated op wrappers into `xent_op_test_base` trigger strict dependency errors.

Furthermore, `xent_op_test_base.py` already imports and tests through the high-level `nn_ops` module:
```python
from tensorflow.python.ops import nn_ops
```
and uses `self._opFwdBwd(...)` (which runs the forward pass and computes the backprop gradient using `backprop.GradientTape` on `nn_ops.softmax_cross_entropy_with_logits`).

### Recommended Fix

1. In `tensorflow/python/kernel_tests/nn_ops/xent_op_test_base.py`:
   - Remove the `from tensorflow.python.ops import gen_nn_ops` import.
   - Update `testDoublePreservesSmallGradient` to use `self._opFwdBwd` or `nn_ops.softmax_cross_entropy_with_logits` (with `backprop.GradientTape`):
   ```python
   @test_util.run_in_graph_and_eager_modes(use_gpu=False)
   def testDoublePreservesSmallGradient(self):
     tail_probability = 5.551115123125776e-17
     _, gradient = self._opFwdBwd(
         labels=np.array([[1.0, 0.0]], dtype=np.float64),
         logits=np.array([[37.42994775023705, 0.0]], dtype=np.float64))
     gradient = self.evaluate(gradient)

     self.assertAllClose(
         [[-tail_probability, tail_probability]], gradient, rtol=1e-14, atol=0)
     self.assertEqual(gradient[0, 0], -gradient[0, 1])
   ```

2. In `tensorflow/python/kernel_tests/nn_ops/BUILD`:
   - Revert the added dependency `"//tensorflow/python/ops:nn_ops_gen"` from `xent_op_test_base`, as `xent_op_test_base` already depends on `"//tensorflow/python/ops:nn_ops"`.

Could you please make these adjustments? Once pushed, the automated test suite will re-run. Thank you!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for your pull request to preserve the float64 dense cross-entropy tail gradients! 

We reviewed your changes, and while the idea of correcting the residual sums is sound, we observed blocking issues in internal compilation and testing that require your attention:

1. **Compilation Error**: The `tensorflow/core/kernels/xent_op.cc` file fails to compile because `<type_traits>` is missing.
   **Fix**: Please add `#include <type_traits>` near the top of the file.

2. **Race Condition in CPU Functor**: The internal CI tests matching your expected results failed (specifically `testDoublePreservesSmallGradient` returned an incorrect `0.0` gradient instead of `-5.55...e-17`). This is because the `XentGradientCorrector<CPUDevice, double>::Correct` method uses `.eval()` on `probabilities`, which executes synchronously on the host caller thread. However, `XentEigenImpl::Compute` schedules the antecedent `backprop` evaluations on Eigen's asynchronous thread pool device (`CPUDevice`). As a result, `.eval()` reads a stale or uninitialized `backprop` buffer before it finishes computing, causing the correction to silently evaluate with garbage values.
   **Fix**: You must redesign the `XentGradientCorrector` to avoid this race condition. Consider removing `.eval()` to keep `probabilities` as a lazy expression, but keep in mind that reducing into `scratch` and then reading `scratch` in a subsequent pointwise device assignment may still lack guaranteed serialization on `CPUDevice` without a barrier. A robust approach is to fold the correction directly into `XentEigenImpl::Compute` or apply a manual loop/synchronization strategy that safely awaits the device's thread pool.

Please push these fixes to your PR branch so we can re-evaluate and merge this improvement!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @GodlyDonuts,

Thank you for your pull request!

I have run the internal presubmit tests on this PR and they succeeded without functional failures. However, our internal C++ style checks flagged that a header include is missing based on "Include What You Use" (IWYU) guidelines:

1. **Missing IWYU Include**: `TTypes` is used for `TTypes<T>::ConstMatrix`, but its header is not explicitly included. Even though the code builds via transitive inclusion, we require all symbols to have their headers directly included.
   Could you please add the corresponding header file near the top of `tensorflow/core/kernels/xent_op.cc`?
   ```cpp
   #include "tensorflow/core/framework/tensor_types.h"
   ```

Please apply this fix, and we will be ready to approve and merge this!

**Review by dmiltr3 — APPROVED** [review]
Thank you for the updates! The code looks good and all tests are passing.

**Review by dmiltr3 — APPROVED** [review]
LGTM!
